File: src/models/activity-report.go
package models

import "github.com/Acsigen/pfa-manager-api/database"

// Contract model, used tags to set the required items
type ActivityReport struct {
	ID          int64
	UserID      int64
	ClientID    int64
	ContractID  int64
	WorkOrderID int64
	InvoiceId   int64
	Name        string  `binding:"required"`
	Date        string  `binding:"required"`
	HoursAmount float64 `binding:"required"`
}

// Add a new contract
func (ar *ActivityReport) Add() error {
	// Build the query, use ? to avoid SQL injection
	query := `INSERT INTO activity_reports(user_id, client_id, contract_id, wo_id, invoice_id, name, date, hours_amount)
	VALUES (?,?,?,?,?,?,?,?)`

	// Prepare the query
	statement, err := database.DB.Prepare(query)
	if err != nil {
		return err
	}

	// Close the statement when the function call ends
	defer statement.Close()

	// Execute the query
	res, err := statement.Exec(ar.UserID, ar.ClientID, ar.ContractID, ar.WorkOrderID, ar.InvoiceId, ar.Name, ar.Date, ar.HoursAmount)
	if err != nil {
		return err
	}

	// Get the last insterted ID
	id, err := res.LastInsertId()

	// Set the ID of the client object so we can return it with the object in another function
	ar.ID = id
	return err
}

// Method for updating a contract
func (ar *ActivityReport) Update() error {
	// Build the query, use ? to avoid SQL injection
	// TODO: Change the invoice ID when invoice section is implemented
	query := `UPDATE activity_reports
	SET invoice_id = 0, name = ?, date = ?, hours_amount = ?
	WHERE id == ?`

	// Prepare the query
	statement, err := database.DB.Prepare(query)
	if err != nil {
		return err
	}

	// Close the statement when the function call ends
	defer statement.Close()

	// Execute the query
	_, err = statement.Exec(ar.Name, ar.Date, ar.HoursAmount, ar.ID)
	return err
}

// Function to get the list of contracts, return a list of contracts and error type
// This is not required to be a method since we don't really use the struct to insert data, we just create a list of contracts
func GetArList(userId int64, clientId int64, contractId int64, woId int64) ([]ActivityReport, error) {
	// Build the query
	query := "SELECT * FROM activity_reports where user_id == ? AND client_id == ? AND contract_id == ? AND wo_id == ?"

	// Direclty execute the query
	rows, err := database.DB.Query(query, userId, clientId, contractId, woId)
	if err != nil {
		return nil, err
	}

	// Close the database connection when function call is done
	defer rows.Close()

	// Create the data structure for the return
	var arList []ActivityReport

	// Iterate over each row
	for rows.Next() {
		// Scan each row and map the items to client properties
		// The order of the arguments for the Scan function must be the same as the DB table not the model
		var ar ActivityReport
		err := rows.Scan(&ar.ID,
			&ar.UserID,
			&ar.ClientID,
			&ar.ContractID,
			&ar.WorkOrderID,
			&ar.InvoiceId,
			&ar.Name,
			&ar.Date,
			&ar.HoursAmount)
		if err != nil {
			return nil, err
		}

		// append each client to the list of clients
		arList = append(arList, ar)
	}

	// Return the clients and no error
	return arList, nil
}

// Function to retrieve the client with a specific ID
// This is not required to be a method since we don't really use the struct to insert data, we just retrieve a client from DB
func GetArById(id int64) (*ActivityReport, error) {
	// We build the query this way to avoid SQL injection
	query := "SELECT * FROM activity_reports WHERE id == ?"
	row := database.DB.QueryRow(query, id)

	// Scan the row and map the items to client properties
	var ar ActivityReport
	err := row.Scan(&ar.ID,
		&ar.UserID,
		&ar.ClientID,
		&ar.ContractID,
		&ar.WorkOrderID,
		&ar.InvoiceId,
		&ar.Name,
		&ar.Date,
		&ar.HoursAmount)
	if err != nil {
		return nil, err
	}

	// Return the client
	return &ar, nil
}

// Method to delete a contract
func (ar ActivityReport) Delete(id int64) error {
	// Query statement
	query := "DELETE FROM activity_report WHERE id == ?"
	statement, err := database.DB.Prepare(query)
	if err != nil {
		return err
	}

	// Close the statement when optimal
	defer statement.Close()

	// execute the query with the id of the contract
	_, err = statement.Exec(id)

	return err
}

File: src/models/activity_report.py
from pydantic import BaseModel
from fastapi import HTTPException
import sqlite3  # For error handling
from ..database import db

class ActivityReport(BaseModel):
    id: int | None = None
    wo_id: int | None = None
    name: str
    date: str
    hours_amount: float
    user_id: int | None = None

    def add(self, client_id: int, contract_id: int, wo_id: int, user_id: int):
        permitted_action: bool = check_permissions(
            client_id=client_id, contract_id=contract_id, wo_id=wo_id, current_user_id=user_id
        )
        if permitted_action:
            query = "INSERT INTO activity_reports(wo_id, name, date, hours_amount, user_id) VALUES (?,?,?,?,?)"
            self.wo_id = wo_id
            self.user_id = user_id
            data: tuple = (
                self.wo_id,
                self.name,
                self.date,
                self.hours_amount,
                self.user_id,
            )
            try:
                res: sqlite3.Cursor = db.execute_query(query=query, params=data)
                self.id = res.lastrowid
                return self
            except sqlite3.Error as e:
                raise HTTPException(500, e.args[0])

    def update(self, client_id: int, contract_id: int, ar_id: int, wo_id: int, user_id: int):
        permitted_action: bool = check_permissions(
            client_id=client_id, contract_id=contract_id, wo_id=wo_id, current_user_id=user_id, ar_id=ar_id  
        )
        if permitted_action:
            query = "UPDATE activity_reports SET name = ?, date = ?, hours_amount = ? WHERE id == ? AND wo_id == ? AND user_id == ?"
            self.id = ar_id
            self.wo_id = wo_id
            self.user_id = user_id
            data = (
                self.name,
                self.date,
                self.hours_amount,
                self.id,
                self.wo_id,
                self.user_id,
            )
            try:
                _: sqlite3.Cursor = db.execute_query(query=query, params=data)
                updated_activity_report: ActivityReport = show_activity_report(
                    activity_report_id=self.id, wo_id=self.wo_id, user_id=self.user_id, client_id=client_id, contract_id=contract_id
                )
                return updated_activity_report
            except sqlite3.Error as e:
                raise HTTPException(500, e.args[0])


def show_activity_report(client_id: int, contract_id: int, wo_id: int, activity_report_id: int, user_id: int):
    permitted_action: bool = check_permissions(
        client_id=client_id, contract_id=contract_id, wo_id=wo_id, current_user_id=user_id, ar_id=activity_report_id
    )
    if permitted_action:
        query = "SELECT * FROM activity_reports WHERE id == ? AND wo_id == ? AND user_id == ?"
        data = (activity_report_id, wo_id, user_id)
        try:
            res: sqlite3.Cursor = db.execute_query(query=query, params=data)
            row: tuple = res.fetchone()
            if row:
                activity_report: ActivityReport = ActivityReport(
                    id=int(row[0]),
                    wo_id=row[1],
                    name=row[2],
                    date=row[3],
                    hours_amount=row[4],
                    user_id=row[5],
                )
                return activity_report
            else:
                raise HTTPException(404, "No such work order")
        except sqlite3.Error as e:
            raise HTTPException(500, e.args[0])


def list_activity_reports(client_id: int, contract_id: int,wo_id: int, user_id: int):
    permitted_action: bool = check_permissions(
            client_id=client_id, contract_id=contract_id, wo_id=wo_id, current_user_id=user_id
        )
    if permitted_action:
        activity_report_list: list[ActivityReport] = []
        query = "SELECT * FROM activity_reports WHERE wo_id == ? AND user_id == ?"
        data: tuple = (wo_id, user_id)
        try:
            res: sqlite3.Cursor = db.execute_query(query=query, params=data)
            rows: list[tuple] = res.fetchall()
            for row in rows:
                activity_report: ActivityReport = ActivityReport(
                    id=int(row[0]),
                    wo_id=row[1],
                    name=row[2],
                    date=row[3],
                    hours_amount=row[4],
                    user_id=row[5],
                )
                activity_report_list.append(activity_report)
            return activity_report_list
        except sqlite3.OperationalError as e:
            raise HTTPException(500, e.args[0])


def delete_activity_report(client_id: int, contract_id: int,wo_id: int, activity_report_id: int, user_id: int):
    permitted_action: bool = check_permissions(
        client_id=client_id, contract_id=contract_id, wo_id=wo_id, current_user_id=user_id, ar_id=activity_report_id
    )
    if permitted_action:
        query = (
            "DELETE FROM activity_reports WHERE id == ? AND wo_id == ? AND user_id == ?"
        )
        data = (activity_report_id, wo_id, user_id)
        try:
            checkout_ar = show_activity_report(
                activity_report_id=activity_report_id, wo_id=wo_id, user_id=user_id, client_id=client_id, contract_id=contract_id
            )
            if type(checkout_ar) is ActivityReport:
                _: sqlite3.Cursor = db.execute_query(query=query, params=data)
                return True
            else:
                raise HTTPException(404, "No such activity_report")
        except sqlite3.IntegrityError as e:
            raise HTTPException(500, e.args[0])

def check_permissions(
    client_id: int, current_user_id: int, contract_id: int, wo_id: int, ar_id: int | None = None
):
    if ar_id:
        query = """SELECT ar.user_id
            FROM activity_reports AS ar
            JOIN work_orders AS wo on ar.wo_id = wo.id
            JOIN contracts AS ctr on wo.contract_id = ctr.id
            JOIN clients AS cli on ctr.client_id = cli.id
            WHERE ctr.id = ? AND cli.id = ? AND wo.id = ? AND ar.id = ?"""
        data = (contract_id, client_id, wo_id, ar_id)
    else:
        query = """SELECT wo.user_id
            FROM work_orders AS wo
            JOIN contracts AS ctr on wo.contract_id = ctr.id
            JOIN clients AS cli on ctr.client_id = cli.id
            WHERE ctr.id = ? AND cli.id = ? AND wo.id = ?"""
        data = (contract_id, client_id, wo_id)
    try:
        res: sqlite3.Cursor = db.execute_query(query=query, params=data)
        row: tuple = res.fetchone()
        if row:
            retrieved_user_id: int = int(row[0])
        else:
            raise HTTPException(
                status_code=403, detail="You are not allowed to perform this action"
            )
    except sqlite3.Error as e:
        raise HTTPException(500, e.args[0])
    if retrieved_user_id == current_user_id:
        return True
    else:
        raise HTTPException(
            status_code=403, detail="You are not allowed to perform this action"
        )